src/ReservoirPriceOracle.sol

```diff
@@ -1,32 +1,405 @@
+    function getQuotes(uint256 aAmount, address aBase, address aQuote)
+        external
+        view
+        validateTokens(aBase, aQuote)
+        returns (uint256 rBidOut, uint256 rAskOut)
+    {
+        uint256 lResult = _getQuote(aAmount, aBase, aQuote);
+        (rBidOut, rAskOut) = (lResult, lResult);
+    }
+
+    // price update related functions
+
+    function gasBountyAvailable() external view returns (uint256) {
+        return address(this).balance;
+    }
+
+    function route(address aToken0, address aToken1) external view returns (address[] memory) {
+        return _route[aToken0][aToken1];
+    }
+
+    /// @notice Updates the TWAP price for all simple routes between `aTokenA` and `aTokenB`. Will also update intermediate routes if the route defined between
+    /// aTokenA and aTokenB is longer than 1 hop
+    /// However, if the route between aTokenA and aTokenB is composite route (more than 1 hop), no cache entry is written
+    /// for priceCache[aTokenA][aTokenB] but instead the prices of its constituent simple routes will be written.
+    /// @param aTokenA Address of one of the tokens for the price update. Does not have to be less than address of aTokenB
+    /// @param aTokenB Address of one of the tokens for the price update. Does not have to be greater than address of aTokenA
+    /// @param aRewardRecipient The beneficiary of the reward. Must implement the receive function if is a smart contract address
+    function updatePrice(address aTokenA, address aTokenB, address aRewardRecipient)
         external
+        validateTokens(aTokenA, aTokenB)
+        nonReentrant
+    {
+        (address lToken0, address lToken1) = aTokenA.sortTokens(aTokenB);
+
+        address[] memory lRoute = _route[lToken0][lToken1];
+        if (lRoute.length == 0) revert PO_NoPath();
+
+        OracleAverageQuery[] memory lQueries = new OracleAverageQuery[](lRoute.length - 1);
+
+        for (uint256 i = 0; i < lRoute.length - 1; ++i) {
+            (lToken0, lToken1) = lRoute[i].sortTokens(lRoute[i + 1]);
+
+            lQueries[i] = OracleAverageQuery(
+                Variable.RAW_PRICE,
+                lToken0,
+                lToken1,
+                twapPeriod,
+                0 // now
+            );
+        }
+
+        uint256[] memory lNewPrices = getTimeWeightedAverage(lQueries);
+
+        for (uint256 i = 0; i < lNewPrices.length; ++i) {
+            address lBase = lQueries[i].base;
+            address lQuote = lQueries[i].quote;
+            uint256 lNewPrice = lNewPrices[i];
+
+            // determine if price has moved beyond the threshold, and pay out reward if so
+            if (_calcPercentageDiff(priceCache[lBase][lQuote], lNewPrice) >= priceDeviationThreshold) {
+                _rewardUpdater(aRewardRecipient);
+            }
+
+            priceCache[lBase][lQuote] = lNewPrice;
+            // TODO: worth the gas cost? who will consume it off chain?
+            emit Price(lBase, lQuote, lNewPrice);
```

Comment:
I think the industry is moving away from events over time, so maybe leave it out for now unless we come up with a strong reason

Reply:
okay removed